File: src/main/java/at/fhhagenberg/sqe/api/ElevatorServiceImpl.java
package at.fhhagenberg.sqe.api;

import at.fhhagenberg.sqe.entity.Direction;
import at.fhhagenberg.sqe.entity.DoorState;
import sqelevator.IElevator;
import at.fhhagenberg.sqe.entity.Elevator;
import com.google.inject.Inject;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.List;

public class ElevatorServiceImpl implements ElevatorService {

    private final IElevator elevatorControl;
    private final ServicedFloorService servicedFloorService;
    private final FloorButtonService floorButtonService;

    @Inject
    public ElevatorServiceImpl(
            IElevator elevatorControl,
            ServicedFloorService servicedFloorService,
            FloorButtonService floorButtonService
    ) {
        this.elevatorControl = elevatorControl;
        this.servicedFloorService = servicedFloorService;
        this.floorButtonService = floorButtonService;
    }

    @Override
    public List<Elevator> getAll() throws RemoteException {
        int totalNumberOfElevators = elevatorControl.getElevatorNum();

        List<Elevator> result = new ArrayList<>(totalNumberOfElevators);
        for (int elevatorNumber = 0; elevatorNumber < totalNumberOfElevators; elevatorNumber++) {
            result.add(get(elevatorNumber));
        }
        return result;
    }

    @Override
    public Elevator get(int elevatorNumber) throws RemoteException {
        int totalNumberOfFloors = elevatorControl.getFloorNum();

        if (elevatorNumber >= 0 && elevatorNumber < totalNumberOfFloors) {
            return new Elevator.Builder()
                    .setElevatorNumber(elevatorNumber)
                    .setAcceleration(elevatorControl.getElevatorAccel(elevatorNumber))
                    .setButtons(floorButtonService.getAll(elevatorNumber))
                    .setCapacity(elevatorControl.getElevatorCapacity(elevatorNumber))
                    .setCommittedDirection(Direction.valueOf(elevatorControl.getCommittedDirection(elevatorNumber)))
                    .setCurrentFloor(elevatorControl.getElevatorFloor(elevatorNumber))
                    .setCurrentPosition(elevatorControl.getElevatorPosition(elevatorNumber))
                    .setCurrentSpeed(elevatorControl.getElevatorSpeed(elevatorNumber))
                    .setCurrentWeight(elevatorControl.getElevatorWeight(elevatorNumber))
                    .setDoorState(DoorState.valueOf(elevatorControl.getElevatorDoorStatus(elevatorNumber)))
                    .setServicedFloors(servicedFloorService.getAll(elevatorNumber))
                    .setTargetFloor(elevatorControl.getTarget(elevatorNumber))
                    .build();
        }
        return null;
    }

    @Override
    public void updateCommittedDirection(Elevator elevator) throws RemoteException {
        elevatorControl.setCommittedDirection(elevator.getElevatorNumber(), elevator.getCommittedDirection().direction);
    }

    @Override
    public void updateTargetFloor(Elevator elevator) throws RemoteException {
        elevatorControl.setTarget(elevator.getElevatorNumber(), elevator.getTargetFloor());
    }
}

File: src/main/kotlin/at/fhhagenberg/sqe/api/ElevatorServiceImpl.kt
package at.fhhagenberg.sqe.api

import at.fhhagenberg.sqe.entity.Direction
import at.fhhagenberg.sqe.entity.DoorState
import at.fhhagenberg.sqe.entity.Elevator
import com.google.inject.Inject
import sqelevator.ConnectableIElevator
import java.rmi.RemoteException
import java.util.*

class ElevatorServiceImpl @Inject constructor(
        private val elevatorControl: ConnectableIElevator,
        private val servicedFloorService: ServicedFloorService,
        private val floorButtonService: FloorButtonService
) : ElevatorService {
    @Throws(RemoteException::class)
    override fun getAll(): List<Elevator> {
        val totalNumberOfElevators = elevatorControl.elevatorNum
        val elevators: MutableList<Elevator> = ArrayList(totalNumberOfElevators.coerceAtLeast(0))
        for (elevatorNumber in 0 until totalNumberOfElevators) {
            get(elevatorNumber)?.let { elevator ->
                elevators.add(elevator)
            }
        }
        return elevators
    }

    @Throws(RemoteException::class)
    override fun get(elevatorNumber: Int): Elevator? {
        return if (elevatorNumber in 0 until elevatorControl.floorNum) {
            Elevator.Builder()
                    .elevatorNumber(elevatorNumber)
                    .acceleration(elevatorControl.getElevatorAccel(elevatorNumber))
                    .buttons(floorButtonService.getAll(elevatorNumber))
                    .capacity(elevatorControl.getElevatorCapacity(elevatorNumber))
                    .committedDirection(Direction.valueOf(elevatorControl.getCommittedDirection(elevatorNumber)))
                    .currentFloor(elevatorControl.getElevatorFloor(elevatorNumber))
                    .currentPosition(elevatorControl.getElevatorPosition(elevatorNumber))
                    .currentSpeed(elevatorControl.getElevatorSpeed(elevatorNumber))
                    .currentWeight(elevatorControl.getElevatorWeight(elevatorNumber))
                    .doorState(DoorState.valueOf(elevatorControl.getElevatorDoorStatus(elevatorNumber)))
                    .servicedFloors(servicedFloorService.getAll(elevatorNumber))
                    .targetFloor(elevatorControl.getTarget(elevatorNumber))
                    .build()
        } else null
    }

    @Throws(RemoteException::class)
    override fun updateTargetFloor(elevator: Elevator, targetFloor: Int) {
        val direction = elevator.evaluateDirection(targetFloor).direction
        elevatorControl.setTarget(elevator.elevatorNumber, targetFloor)
        elevatorControl.setCommittedDirection(elevator.elevatorNumber, direction)
    }
}